### PR TITLE
Bump some libs version for dev3 branch

### PR DESCRIPTION
**Pull request by meiMingle**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -57,22 +57,25 @@ subprojects {
...
-        def jpVersion = '3.25.4'
+        def jadxVersion = '1.4.7'
+        def jpVersion = '3.25.5'
         def junitVersion = '5.10.0'
-        def lljzipVersion = '2.1.3'
+        def lljzipVersion = '2.3.0'
         def ssvmVersion = '1.9.0.3'
-        def quiltflowerVersion = '1.9.0'
+        // def quiltflowerVersion = '1.9.0'
+        def vineflowerVersion = '1.9.3'
```
Comment (Col-E) [inline build.gradle:75]: The commented out value can be removed

File: build.gradle [hunk 2]
```diff
@@ -102,10 +105,12 @@ subprojects {
         junit_engine = "org.junit.jupiter:junit-jupiter-engine:$junitVersion"
         junit_params = "org.junit.jupiter:junit-jupiter-params:$junitVersion"
         llzip = "software.coley:lljzip:$lljzipVersion"
-        logback_classic = 'ch.qos.logback:logback-classic:1.2.10' // newer releases break in jar releases
-        pe = 'com.github.cademtz:JavaPeParser:1.0.0' // fails on 2.2.3, breaking changes
+        slf4j = "org.slf4j:slf4j-api:$slf4jVersion"
+        logback_classic = "ch.qos.logback:logback-classic:$logbackVersion" // quiltflower -> vineflower
```
Comment (Col-E) [inline build.gradle:109]: The comment on qf --> vf is on the wrong line

File: build.gradle [hunk 3]
```diff
@@ -102,10 +105,12 @@ subprojects {
...
+        pe = 'com.github.cademtz:JavaPeParser:2.2.3'
         picocli = 'info.picocli:picocli:4.7.4'
-        quiltflower = "org.quiltmc:quiltflower:$quiltflowerVersion"
+        // quiltflower = "org.quiltmc:quiltflower:$quiltflowerVersion"
```
Comment (Col-E) [inline build.gradle:112]: Commented out line can be removed

File: recaf-core/src/main/java/me/coley/recaf/decompile/quilt/QuiltFlowerDecompiler.java [hunk 4]
```diff
@@ -17,12 +17,13 @@
...
 import java.io.ByteArrayInputStream;
 import java.io.InputStream;
+import java.net.URL;
 import java.util.*;
 import java.util.jar.Manifest;
 import java.util.stream.Collectors;
 
 /**
- * QuiltFlower decompiler implementation.
+ * VineFlower decompiler implementation.
```
Comment (Col-E) [inline recaf-core/src/main/java/me/coley/recaf/decompile/quilt/QuiltFlowerDecompiler.java:26]: Class names and package should be updated as well

File: recaf-core/src/main/java/me/coley/recaf/decompile/quilt/QuiltFlowerDecompiler.java [hunk 5]
```diff
@@ -35,9 +36,34 @@ public class QuiltFlowerDecompiler extends Decompiler {
...
-		super("QuiltFlower", "1.9.0");
+		super("VineFlower", findQuiltFlowerVersion());
 		fernFlowerProperties.put("ind", "    ");
 	}
+    private static String findQuiltFlowerVersion() {
+        try {
+            ClassLoader classLoader = Fernflower.class.getClassLoader();
+            if (classLoader != null) {
+                Enumeration<URL> resources = classLoader.getResources("META-INF/MANIFEST.MF");
+				while (resources.hasMoreElements()) {
```
Comment (Col-E) [inline recaf-core/src/main/java/me/coley/recaf/decompile/quilt/QuiltFlowerDecompiler.java:47]: This won't work in JAR release form. The reason why this works in the IDE is because each dependency is loaded from its own JAR file. If we combine the jars, and replace things like the `MANIFEST.MF` with our own then the information this relies on will no longer exist.

File: recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/DosTableDisplayMode.java [hunk 6]
```diff
@@ -2,17 +2,19 @@
 
 import me.coley.recaf.ui.pane.table.SizedDataTypeTable;
 import me.coley.recaf.ui.pane.table.TableDisplayMode;
-import me.martinez.pe.ImageDosHeader;
-import me.martinez.pe.ImagePeHeaders;
+// import me.martinez.pe.ImageDosHeader;
+// import me.martinez.pe.ImagePeHeaders;
+import me.martinez.pe.PeImage;
```
Comment (Col-E) [inline recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/DosTableDisplayMode.java:7]: Commented out lines can be removed.

File: recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/ExportTableDisplayMode.java [hunk 7]
```diff
@@ -2,17 +2,19 @@
 
 import me.coley.recaf.ui.pane.table.SizedDataTypeTable;
 import me.coley.recaf.ui.pane.table.TableDisplayMode;
-import me.martinez.pe.CachedExportEntry;
-import me.martinez.pe.CachedImageExports;
+import me.martinez.pe.ExportEntry;
+import me.martinez.pe.LibraryExport;
+// import me.martinez.pe.CachedExportEntry;
+// import me.martinez.pe.CachedImageExports;
```
Comment (Col-E) [inline recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/ExportTableDisplayMode.java:8]: Commented out lines again.

File: recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/FileTableDisplayMode.java [hunk 8]
```diff
@@ -3,8 +3,10 @@
 import me.coley.recaf.ui.pane.table.SizedDataTypeTable;
 import me.coley.recaf.ui.pane.table.TableDisplayMode;
 import me.coley.recaf.ui.pane.table.TableWord;
-import me.martinez.pe.ImageFileHeader;
-import me.martinez.pe.ImagePeHeaders;
+import me.martinez.pe.PeImage;
+import me.martinez.pe.headers.ImageFileHeader;
+// import me.martinez.pe.ImageFileHeader;
+// import me.martinez.pe.ImagePeHeaders;
```
Comment (Col-E) [inline recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/FileTableDisplayMode.java:9]: Commented out lines again.

File: recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/ImportTableDisplayMode.java [hunk 9]
```diff
@@ -2,17 +2,19 @@
 
 import me.coley.recaf.ui.pane.table.SizedDataTypeTable;
 import me.coley.recaf.ui.pane.table.TableDisplayMode;
-import me.martinez.pe.CachedImportEntry;
-import me.martinez.pe.CachedLibraryImports;
+import me.martinez.pe.ExportEntry;
+import me.martinez.pe.LibraryExport;
+// import me.martinez.pe.CachedImportEntry;
+// import me.martinez.pe.CachedLibraryImports;
```
Comment (Col-E) [inline recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/ImportTableDisplayMode.java:8]: Commented out lines again.

File: recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/OptionalTableDisplayMode.java [hunk 10]
```diff
@@ -3,8 +3,10 @@
 import me.coley.recaf.ui.pane.table.SizedDataTypeTable;
 import me.coley.recaf.ui.pane.table.TableDisplayMode;
 import me.coley.recaf.ui.pane.table.TableWord;
-import me.martinez.pe.ImageOptionalHeader;
-import me.martinez.pe.ImagePeHeaders;
+// import me.martinez.pe.ImageOptionalHeader;
+// import me.martinez.pe.ImagePeHeaders;
```
Comment (Col-E) [inline recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/OptionalTableDisplayMode.java:7]: Commented out lines again.

File: recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/PEExplorerPane.java [hunk 11]
```diff
@@ -14,13 +14,20 @@
 import me.coley.recaf.ui.pane.table.SizedDataTypeTable;
 import me.coley.recaf.ui.pane.table.TableGeneric;
 import me.coley.recaf.util.logging.Logging;
-import me.martinez.pe.CachedLibraryImports;
-import me.martinez.pe.ImagePeHeaders;
-import me.martinez.pe.ImageSectionHeader;
+// import me.martinez.pe.CachedLibraryImports;
+// import me.martinez.pe.ImagePeHeaders;
+// import me.martinez.pe.ImageSectionHeader;
```
Comment (Col-E) [inline recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/PEExplorerPane.java:19]: Commented out lines again.

File: recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/SectionTableDisplayMode.java [hunk 12]
```diff
@@ -2,7 +2,8 @@
 
 import me.coley.recaf.ui.pane.table.SizedDataTypeTable;
 import me.coley.recaf.ui.pane.table.TableDisplayMode;
-import me.martinez.pe.ImageSectionHeader;
+import me.martinez.pe.headers.ImageSectionHeader;
+// import me.martinez.pe.ImageSectionHeader;
```
Comment (Col-E) [inline recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/SectionTableDisplayMode.java:6]: Commented out lines again.

File: recaf-ui/src/main/java/me/coley/recaf/ui/pane/table/SizedDataTypeTable.java [hunk 13]
```diff
@@ -1,7 +1,8 @@
 package me.coley.recaf.ui.pane.table;
 
 import javafx.scene.control.TableView;
-import me.martinez.pe.ImagePeHeaders;
+// import me.martinez.pe.ImagePeHeaders;
```
Comment (Col-E) [inline recaf-ui/src/main/java/me/coley/recaf/ui/pane/table/SizedDataTypeTable.java:4]: Commented out lines again.

File: recaf-ui/src/main/java/me/coley/recaf/ui/pane/table/SizedDataTypeTable.java [hunk 14]
```diff
@@ -134,7 +135,8 @@ public void addQword(String memberName, String value, String meaning) {
 	 * @param pe
 	 * 		PE image to pull 64-bit flag from.
 	 */
-	public void addAddress(String memberName, long value, String meaning, ImagePeHeaders pe) {
+	// public void addAddress(String memberName, long value, String meaning, ImagePeHeaders pe) {
```
Comment (Col-E) [inline recaf-ui/src/main/java/me/coley/recaf/ui/pane/table/SizedDataTypeTable.java:138]: Commented out lines again.

**Review by Col-E — COMMENTED** [review]
Thank you for the updates 👍 

Couple of things like commented out code to address and renaming related to QF -> VF.

**Comment by meiMingle** [comment]
@Col-E I have completed the modifications as required, please review again

File: recaf-core/README.md [hunk 15]
```diff
@@ -25,7 +25,7 @@ and the basis to create UI presentations off of.
   - `fallback` is a disassembling decompiler intended for usage when all others fail. 
   - `jadx` is a standard decompiler designed for Android but supports standard Java decompilation, from [skylot/jadx](https://github.com/skylot/jadx)
   - `procyon` is a standard decompiler, from [mstrobel/procyon](https://github.com/mstrobel/procyon)
-  - `quilt` is a standard decompiler forked from JetBrain's FernFlower, from [QuiltMC/quiltflower](https://github.com/QuiltMC/quiltflower)
+  - `vine` is a standard decompiler forked from JetBrain's FernFlower, from [Vineflower/vineflower](https://github.com/Vineflower/vineflower).It used to be called QuiltFlower in version 1.9.0 and before.
```
Comment (Col-E) [inline recaf-core/README.md:28]: The last sentence isn't super necessary.

File: build.gradle [hunk 16]
```diff
@@ -102,10 +104,11 @@ subprojects {
         junit_engine = "org.junit.jupiter:junit-jupiter-engine:$junitVersion"
         junit_params = "org.junit.jupiter:junit-jupiter-params:$junitVersion"
         llzip = "software.coley:lljzip:$lljzipVersion"
-        logback_classic = 'ch.qos.logback:logback-classic:1.2.10' // newer releases break in jar releases
-        pe = 'com.github.cademtz:JavaPeParser:1.0.0' // fails on 2.2.3, breaking changes
+        slf4j = "org.slf4j:slf4j-api:$slf4jVersion"
+        logback_classic = "ch.qos.logback:logback-classic:$logbackVersion"
```
Comment (Col-E) [inline build.gradle:108]: The comment for logback-classic still applies and should not be updated.

Running `gradlew build` and attempting to run the created jar fails:
```
SLF4J: A SLF4J service provider failed to instantiate:
org.slf4j.spi.SLF4JServiceProvider: Provider ch.qos.logback.classic.spi.LogbackServiceProvider not found
SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See https://www.slf4j.org/codes.html#noProviders for further details.
Exception in thread "main" java.lang.ClassCastException: class org.slf4j.helpers.NOPLoggerFactory cannot be cast to class ch.qos.logback.classic.LoggerContext (org.slf4j.helpers.NOPLoggerFactory and ch.qos.logback.classic.LoggerContext are in unnamed module of loader 'app')
        at me.coley.recaf.util.logging.Logging.addFileAppender(Logging.java:83)
        at me.coley.recaf.Main.setupLogging(Main.java:85)
        at me.coley.recaf.Main.main(Main.java:39)
        ```

Comment (meiMingle) [reply build.gradle:108]: image
This problem should be caused by the mismatch between slf4j and logback versions. This problem no longer exists. You can clone my code and try it.
```bash
git clone -b dev3-optimize --single-branch https://github.com/meiMingle/Recaf.git Recaf
cd Recaf
gradle build
```

Comment (meiMingle) [reply build.gradle:108]: Now it's fixed

**Comment by meiMingle** [comment]
@Col-E Finished，review again pls

File: recaf-core/README.md [hunk 17]
```diff
@@ -25,7 +25,7 @@ and the basis to create UI presentations off of.
   - `fallback` is a disassembling decompiler intended for usage when all others fail. 
   - `jadx` is a standard decompiler designed for Android but supports standard Java decompilation, from [skylot/jadx](https://github.com/skylot/jadx)
   - `procyon` is a standard decompiler, from [mstrobel/procyon](https://github.com/mstrobel/procyon)
-  - `vine` is a standard decompiler forked from JetBrain's FernFlower, from [Vineflower/vineflower](https://github.com/Vineflower/vineflower).It used to be called QuiltFlower in version 1.9.0 and before.
```
Comment (Amejonah1200) [inline recaf-core/README.md]: Why did you remove it? For ppl who don't know about the change, it is good to provide that info.

Comment (Col-E) [reply recaf-core/README.md]: I don't think the last sentence is relevant, see my prior comments.

**Comment by Col-E** [comment]
Looks like the import/export table behavior changed.

Old behavior:
image

New behavior:
image

Sample:
[natives.zip](https://github.com/Col-E/Recaf/files/12885959/natives.zip)

@cademtz Any input?

File: recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/PEExplorerPane.java [hunk 18]
```diff
@@ -91,28 +95,34 @@ public void onUpdate(FileInfo newValue) {
...
-			CachedLibraryImports cachedLibraryImport = pe.getCachedLibraryImport(i);
-			TreeItem<String> libraryItem = new TreeItem<>(cachedLibraryImport.getName());
-			itemImportDirectory.getChildren().add(libraryItem);
+		if (pe.exports.isOk()) {
+			LibraryExport libraryExport = pe.exports.getOk();
+			for (int i = 0; i < libraryExport.entries.length; i++) {
+				ExportEntry exportEntry = libraryExport.entries[i];
+				TreeItem<String> libraryItem = new TreeItem<>(exportEntry.name);
+				itemImportDirectory.getChildren().add(libraryItem);
+			}
```
Comment (Col-E) [inline recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/PEExplorerPane.java:122]: The diff in this section seems to change behavior. I've not dug into the new version of JavaPeParser to find the equivalent new behavior, but at a glance this seems to swap imports and exports.

File: recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/PEExplorerPane.java [hunk 19]
```diff
@@ -180,19 +190,19 @@ private void onSelectionChange(ObservableValue<? extends TreeItem<String>> obser
...
 
 			if (sectionIndex != -1) {
-				ImageSectionHeader sectionHeader = pe.sectionHeaders[sectionIndex];
+				ImageSectionHeader sectionHeader = pe.sectionHeaders.get(sectionIndex).getOk();
 				SECTION_MODE.apply(sectionHeader, primaryTableView);
 			} else if (importIndex != -1) {
-				CachedLibraryImports importEntry = pe.getCachedLibraryImport(importIndex);
-				IMPORT_MODE.apply(importEntry, primaryTableView);
+				LibraryExport libraryExport = pe.exports.getOk();
+				IMPORT_MODE.apply(libraryExport, primaryTableView);
```
Comment (Col-E) [inline recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/PEExplorerPane.java:205]: Similarly here, import/export swapped.

File: recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/PEExplorerPane.java (hunk 18 above)
Comment (cademtz) [reply recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/PEExplorerPane.java:122]: Yeah, the new line no. 122 is adding `exportEntry.name` to the control `itemImportDirectory`. Seems like a simple hiccup in the refactoring.

Aside from slight renaming, The only changed behavior in my new API is that many values are replaced with a `Result` class, and you have to call `getOk()` or `getOkOrDefault(null)`.
*(You can also print/display `peImage.warnings` to see comments from the parser)*

**Comment by cademtz** [comment]
On top of imports/exports being swapped, it looks like the collapsible "Import directory" on the left isn't showing library names. It's showing the entries contained in those libraries instead.

```java
ArrayList<LibraryImports> libs;
```
This type is basically a list of DLL files. So you need only to iterate and add `libs.get(i).name` to that collapsible list.

File: recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/PEExplorerPane.java (hunk 18 above)
Comment (meiMingle) [reply recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/PEExplorerPane.java:124]: Fixed

File: recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/PEExplorerPane.java (hunk 19 above)
Comment (meiMingle) [reply recaf-ui/src/main/java/me/coley/recaf/ui/pane/pe/PEExplorerPane.java]: Fixed

**Comment by meiMingle** [comment]
Completed

**Comment by Col-E** [comment]
Thank you for the updates and your patience, been quite busy lately 😅
